# Exercises 3. Introduction to data analysis

## Exercise 3.1

### Problem 1 - Read the file and clean it

Import the pandas module and read the weather data into a variable called data. Print the first five rows of the data file.

In [1]:
import pandas
data_source = pandas.read_csv("6153237444115dat.csv")
data_source.head()

USAF   WBAN  YR--MODAHRMN  DIR SPD  GUS  CLG  SKC  L  M  ...     SLP  \
0  28450  99999  201705010000  174  10   14  ***  ***  *  *  ...  1009.2   
1  28450  99999  201705010020  180  10  ***    4  ***  *  *  ...  ******   
2  28450  99999  201705010050  190  10  ***    4  ***  *  *  ...  ******   
3  28450  99999  201705010100  188  12   16  ***  ***  *  *  ...  1009.1   
4  28450  99999  201705010120  200  13  ***    2  OBS  *  *  ...  ******   

     ALT     STP  MAX  MIN  PCP01  PCP06  PCP24  PCPXX  SD  
0  *****   984.1  ***  ***  *****  *****  *****  *****  35  
1  29.74  ******  ***  ***  *****  *****  *****  *****  **  
2  29.74  ******  ***  ***  *****  *****  *****  *****  **  
3  *****   984.0  ***  ***  *****  *****  *****  *****  35  
4  29.74  ******  ***  ***  *****  *****  *****  *****  **  

[5 rows x 33 columns]

### Problem 2 - Basic characteristics of the data

Based on the data DataFrame from Problem 1, answer to following questions:
1. How many rows is there in the data?
2. What are the column names?
3. What are the datatypes of the columns?

In [2]:
print(len(data_source))
print(data_source.columns.values)
print(data_source.dtypes)

11694
['USAF' 'WBAN' 'YR--MODAHRMN' 'DIR' 'SPD' 'GUS' 'CLG' 'SKC' 'L' 'M' 'H'
 'VSB' 'MW' 'MW.1' 'MW.2' 'MW.3' 'AW' 'AW.1' 'AW.2' 'AW.3' 'W' 'TEMP'
 'DEWP' 'SLP' 'ALT' 'STP' 'MAX' 'MIN' 'PCP01' 'PCP06' 'PCP24' 'PCPXX' 'SD']
USAF             int64
WBAN             int64
YR--MODAHRMN     int64
DIR             object
SPD             object
GUS             object
CLG             object
SKC             object
L               object
M               object
H               object
VSB             object
MW              object
MW.1            object
MW.2            object
MW.3            object
AW              object
AW.1            object
AW.2            object
AW.3            object
W               object
TEMP            object
DEWP            object
SLP             object
ALT             object
STP             object
MAX             object
MIN             object
PCP01           object
PCP06           object
PCP24           object
PCPXX           object
SD              object
dtype: object


### Problem 3 - Descriptive statistics

Based on the data DataFrame from Problem 1, answer to following questions:
 - What is the mean Fahrenheit temperature in the data (use the TEMP column)?
 - What is the standard deviation of the Maximum temperature (use the MAX column)?
 - How many unique stations exists in the data (use the USAF column)?

In [3]:
# Операции mean и std выдали ошибки, поэтому 
print(data_source['TEMP'].unique())
print(data_source['MAX'].unique())

['31' '30' '32' '34' '36' '37' '39' '40' '41' '42' '43' '44' '38' '45'
 '35' '33' '46' '47' '48' '49' '50' '51' '52' '53' '54' '28' '29' '27'
 '25' '26' '23' '24' '21' '19' '20' '****' '55' '56' '57' '59' '60' '63'
 '62' '64' '65' '66' '67' '68' '69' '70' '61' '58' '71' '72' '73' '74'
 '75' '77']
['***' '34' '44' '42' '45' '37' '39' '48' '54' '41' '35' '33' '29' '31'
 '32' '38' '36' '40' '46' '53' '50' '52' '55' '59' '49' '43' '51' '57'
 '69' '73' '71' '74' '70' '65' '58' '63' '60' '64' '68' '61' '47' '62'
 '67' '72' '66' '56' '75' '76' '77']


In [4]:
data_source_corrected = pandas.read_csv("6153237444115dat.csv", na_values=["*", "**", "***", "****", "*****", "******"])
data_source_corrected.head()

USAF   WBAN  YR--MODAHRMN    DIR   SPD   GUS  CLG  SKC   L   M  ...  \
0  28450  99999  201705010000  174.0  10.0  14.0  NaN  NaN NaN NaN  ...   
1  28450  99999  201705010020  180.0  10.0   NaN  4.0  NaN NaN NaN  ...   
2  28450  99999  201705010050  190.0  10.0   NaN  4.0  NaN NaN NaN  ...   
3  28450  99999  201705010100  188.0  12.0  16.0  NaN  NaN NaN NaN  ...   
4  28450  99999  201705010120  200.0  13.0   NaN  2.0  OBS NaN NaN  ...   

      SLP    ALT    STP  MAX  MIN  PCP01  PCP06  PCP24  PCPXX    SD  
0  1009.2    NaN  984.1  NaN  NaN    NaN    NaN    NaN    NaN  35.0  
1     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  
2     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  
3  1009.1    NaN  984.0  NaN  NaN    NaN    NaN    NaN    NaN  35.0  
4     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  

[5 rows x 33 columns]

In [5]:
workdata = data_source_corrected.copy()
print(workdata.dtypes)

USAF              int64
WBAN              int64
YR--MODAHRMN      int64
DIR             float64
SPD             float64
GUS             float64
CLG             float64
SKC              object
L               float64
M               float64
H               float64
VSB             float64
MW              float64
MW.1            float64
MW.2            float64
MW.3            float64
AW              float64
AW.1            float64
AW.2            float64
AW.3            float64
W               float64
TEMP            float64
DEWP            float64
SLP             float64
ALT             float64
STP             float64
MAX             float64
MIN             float64
PCP01           float64
PCP06           float64
PCP24           float64
PCPXX           float64
SD              float64
dtype: object


In [6]:
print("Mean Fahrenheit temperature in the data is", workdata["TEMP"].mean())
print("Standard deviation of the Maximum temperature is", workdata["MAX"].std())
print(workdata["USAF"].unique(),"- numbers of unique stations existing in the data")

Mean Fahrenheit temperature in the data is 52.24796852279531
Standard deviation of the Maximum temperature is 10.30798909464196
[28450 29980] - numbers of unique stations existing in the data


## Exercise 3.2

In this exercise, you will clean the data from our data file by removing no-data values, convert temperature values in Fahrenheit to Celsius, and split the data into separate datasets using the weather station identification code. We will start this problem by cleaning and converting our temperature data. An overview of the tasks in this exercise:
 - Create a new dataframe called selected that contains selected columns from the data file
 - Clean the new DataFrame by removing no-data values
 - Create a new column for temperatures converted from Fahrenheit to Celsius
 - Divide the data into separate DataFrames for the Helsinki Kumpula and Rovaniemi stations
 - Save the new DataFrames to CSV files

## Problem 1 - Read the data and remove NaN values

The first step for this problem is to read the data file 6153237444115dat.csv into a variable data using pandas and cleaning it a bit:
 - Select the columns USAF, YR--MODAHRMN, TEMP, MAX, MIN from the data DataFrame and assign them to a variable selected
 - Remove all rows from selected that have NoData in the column TEMP using the dropna() function

In [7]:
workdata_cropped = workdata[["USAF", "YR--MODAHRMN", "TEMP", "MAX", "MIN"]].copy()
workdata_cropped = workdata_cropped.dropna(subset=["TEMP"])
print("Is nans in TEMP? - ", workdata_cropped["TEMP"].hasnans)
workdata_cropped.head()

Is nans in TEMP? -  False


USAF  YR--MODAHRMN  TEMP  MAX  MIN
0  28450  201705010000  31.0  NaN  NaN
1  28450  201705010020  30.0  NaN  NaN
2  28450  201705010050  30.0  NaN  NaN
3  28450  201705010100  31.0  NaN  NaN
4  28450  201705010120  30.0  NaN  NaN

### Problem 2 - Convert temperatures to Celsius

Convert the temperature values from Fahrenheits to Celsius:
 - Create a new column to selected called Celsius.
 - Convert the Fahrenheit temperatures from TEMP using the conversion formula below and store the results in the new Celsius column:
   - TempCelsius = (TempFahrenheit - 32) / 1.8
 - Round the values in the Celsius column to have 0 decimals (do not create a new column, update the current one)
 - Convert the Celsius values into integers (do not create a new column, update the current one)

In [8]:
workdata_cropped["CELSIUS"] = 0.0
workdata_cropped["CELSIUS"] = (workdata_cropped["TEMP"] - 32) / 1.8
workdata_cropped["CELSIUS"] = workdata_cropped["CELSIUS"].round()
workdata_cropped["CELSIUS"] = workdata_cropped["CELSIUS"].astype(int)
print("CELSIUS type is", workdata_cropped["CELSIUS"].dtype)
workdata_cropped.head()


# Если формула была бы более сложной, то следовало бы воспользоваться apply()

CELSIUS type is int32


USAF  YR--MODAHRMN  TEMP  MAX  MIN  CELSIUS
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1

### Problem 3 - Select data and save to disk

Divide the data in selected into two separate DataFrames:
 - Select all rows from the selected DataFrame with the USAF code 29980 into a variable called kumpula.
 - Select all rows from the selected DataFrame with the USAF code 28450 into a variable called rovaniemi.
 - Save the kumpula DataFrame into a file Kumpula_temps_May_Aug_2017.csv in CSV format:
   - Separate the columns with commas (,)
   - Use only 2 decimals for the floating point numbers
 - Repeat the same procedures and save the rovaniemi DataFrame into a file Rovaniemi_temps_May_Aug_2017.csv.

In [9]:
kumpula = workdata_cropped[workdata_cropped["USAF"] == 29980]
print(kumpula.head)
rovaniemi = workdata_cropped[workdata_cropped["USAF"] == 28450]
print(rovaniemi.head)
kumpula.to_csv("Kumpula_temps_May_Aug_2017.csv", sep=',', float_format="%.2f")
rovaniemi.to_csv("Rovaniemi_temps_May_Aug_2017.csv", sep=',', float_format="%.2f")

# pandas ниже отбросил один ноль для столбца TEMP, но в файле их два

<bound method NDFrame.head of         USAF  YR--MODAHRMN  TEMP  MAX  MIN  CELSIUS
8770   29980  201705010000  37.0  NaN  NaN        3
8771   29980  201705010100  37.0  NaN  NaN        3
8772   29980  201705010200  37.0  NaN  NaN        3
8773   29980  201705010300  37.0  NaN  NaN        3
8774   29980  201705010400  39.0  NaN  NaN        4
...      ...           ...   ...  ...  ...      ...
11689  29980  201708311900  64.0  NaN  NaN       18
11690  29980  201708312000  64.0  NaN  NaN       18
11691  29980  201708312100  64.0  NaN  NaN       18
11692  29980  201708312200  64.0  NaN  NaN       18
11693  29980  201708312300  64.0  NaN  NaN       18

[2924 rows x 6 columns]>
<bound method NDFrame.head of        USAF  YR--MODAHRMN  TEMP  MAX  MIN  CELSIUS
0     28450  201705010000  31.0  NaN  NaN       -1
1     28450  201705010020  30.0  NaN  NaN       -1
2     28450  201705010050  30.0  NaN  NaN       -1
3     28450  201705010100  31.0  NaN  NaN       -1
4     28450  201705010120  30.0  Na

## Exercise 3.3

In this Exercise, we will explore our temperature data by comparing spring temperatures between Kumpula and Rovaniemi. To do this we’ll use some conditions to extract subsets of our data and then analyse these subsets using basic pandas functions. Notice that in this exercise, we will use data saved from the previous Exercise (2.2.6), hence you should finish that Exercise before this one. An overview of the tasks in this exercise:
 - Calculate the median temperatures for Kumpula and Rovaniemi for the summer of 2017
 - Select temperatures for May and June 2017 in separate DataFrames for each location
 - Calculate descriptive statistics for each month (May, June) and location (Kumpula, Rovaniemi)

### Problem 1 - Read the data and calculate basic statistics


Read in the CSV files generated in Exercise 2.2.6 to the variables kumpula and rovaniemi and answer to following questions:
 - What was the median Celsius temperature during the observed period in Helsinki Kumpula? Store the answer in a variable kumpula_median.
 - What was the median Celsius temperature during the observed period in Rovaniemi? Store the answer in a variable rovaniemi_median.

In [10]:
kumpula_data = pandas.read_csv("Kumpula_temps_May_Aug_2017.csv", index_col=0)
rovaniemi_data = pandas.read_csv("Rovaniemi_temps_May_Aug_2017.csv", index_col=0)
print(kumpula_data["CELSIUS"].median())
print(rovaniemi_data["CELSIUS"].median())

14.0
11.0


In [11]:
kumpula_data.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN  CELSIUS
8770  29980  201705010000  37.0  NaN  NaN        3
8771  29980  201705010100  37.0  NaN  NaN        3
8772  29980  201705010200  37.0  NaN  NaN        3
8773  29980  201705010300  37.0  NaN  NaN        3
8774  29980  201705010400  39.0  NaN  NaN        4

In [12]:
rovaniemi_data.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN  CELSIUS
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1

### Problem 2 - Select data and compare temperatures between months

The median temperatures above consider data from the entire summer (May-Aug), hence the differences might not be so clear. Let’s now find out the mean temperatures from May and June 2017 in Kumpula and Rovaniemi:
 - From the kumpula and rovaniemi DataFrames, select the rows where values of the YR--MODAHRMN column are from May 2017. Assign these selected rows into the variables kumpula_may and rovaniemi_may
 - Repeat the procedure for the month of June and assign those values into variables to kumpula_june and rovaniemi_june
 - Calculate and print the mean, min and max Celsius temperatures for both places in May and June using the new subset dataframes (kumpula_may, rovaniemi_may, kumpula_june, and rovaniemi_june). Answer to following questions:
   - Does there seem to be a large difference in temperatures between the months?
   - Is Rovaniemi a much colder place than Kumpula?

In [13]:
kumpula_data["YR--MODAHRMN"] = pandas.to_datetime(kumpula_data["YR--MODAHRMN"], format='%Y%m%d%H%M')
rovaniemi_data["YR--MODAHRMN"] = pandas.to_datetime(rovaniemi_data["YR--MODAHRMN"], format='%Y%m%d%H%M')
kumpula_may = kumpula_data[kumpula_data["YR--MODAHRMN"].dt.month == 5]
rovaniemi_may = rovaniemi_data[rovaniemi_data["YR--MODAHRMN"].dt.month == 5]
kumpula_june = kumpula_data[kumpula_data["YR--MODAHRMN"].dt.month == 6]
rovaniemi_june = rovaniemi_data[rovaniemi_data["YR--MODAHRMN"].dt.month == 6]
print("Mean Celsius temperatures in Kumpula in May 2017", kumpula_may["CELSIUS"].mean())
print("Min Celsius temperatures in Kumpula in May 2017",  kumpula_may["CELSIUS"].min())
print("Max Celsius temperatures in Kumpula in May 2017",  kumpula_may["CELSIUS"].max())
print()
print("Mean Celsius temperatures in Rovaniemi in May 2017", rovaniemi_may["CELSIUS"].mean())
print("Min Celsius temperatures in Rovaniemi in May 2017",  rovaniemi_may["CELSIUS"].min())
print("Max Celsius temperatures in Rovaniemi in May 2017",  rovaniemi_may["CELSIUS"].max())
print()
print("Mean Celsius temperatures in Kumpula in June 2017", kumpula_june["CELSIUS"].mean())
print("Min Celsius temperatures in Kumpula in June 2017",  kumpula_june["CELSIUS"].min())
print("Max Celsius temperatures in Kumpula in June 2017",  kumpula_june["CELSIUS"].max())
print()
print("Mean Celsius temperatures in Rovaniemi in June 2017", rovaniemi_june["CELSIUS"].mean())
print("Min Celsius temperatures in Rovaniemi in June 2017",  rovaniemi_june["CELSIUS"].min())
print("Max Celsius temperatures in Rovaniemi in June 2017",  rovaniemi_june["CELSIUS"].max())

Mean Celsius temperatures in Kumpula in May 2017 9.758434547908232
Min Celsius temperatures in Kumpula in May 2017 -2
Max Celsius temperatures in Kumpula in May 2017 23

Mean Celsius temperatures in Rovaniemi in May 2017 2.986036036036036
Min Celsius temperatures in Rovaniemi in May 2017 -7
Max Celsius temperatures in Rovaniemi in May 2017 15

Mean Celsius temperatures in Kumpula in June 2017 13.732492997198879
Min Celsius temperatures in Kumpula in June 2017 3
Max Celsius temperatures in Kumpula in June 2017 24

Mean Celsius temperatures in Rovaniemi in June 2017 11.017395392571697
Min Celsius temperatures in Rovaniemi in June 2017 -1
Max Celsius temperatures in Rovaniemi in June 2017 23


### Problem 3 - Parse daily temperatures by aggregating data

In this problem, the aim is to aggregate the hourly temperature data for Kumpula and Rovaniemi weather stations to a daily level. Currently, there are at most three measurements per hour in the data, as you can see from the YR–MODAHRMN column:

|   | USAF  | YR--MODAHRMN | TEMP | MAX | MIN | CELSIUS |
|---|-------|--------------|------|-----|-----|---------|
| 0 | 28450 | 201705010000 | 31.0 | NaN | NaN |   -1    |
| 1 | 28450 | 201705010020 | 30.0 | NaN | NaN |   -1    |
| 2 | 28450 | 201705010050 | 30.0 | NaN | NaN |   -1    |
| 3 | 28450 | 201705010100 | 31.0 | NaN | NaN |   -1    |
| 4 | 28450 | 201705010120 | 30.0 | NaN | NaN |   -1    |

In this problem you should:
 - Summarize the information for each day by aggregating (grouping) the DataFrame using the groupby() function.
 - The output should be a new DataFrame where you have calculated mean, max and min Celsius temperatures for each day separately based on hourly values.
 - Repeat the task for the two data sets you created in Problem 2 (May-August temperatures from Rovaniemi and Kumpula).

In [14]:
# Я не использую resample() вместо groupby(), потому что в предыдущей же буквально подглаве рассказывалось 
# как работать с datetime и выбирать данные с его помощью. 
# Поэтому я не знаю почему в задании расмматривается вариант использования даты как string

# Также я не понимаю для чего нам агрегировать на основе часовых значений, если можно сразу взять за день

# Значения по фаренгейту и по цельсию не совпадают, так как мы изначально берём округлённые до целого данные,
# см. Exercise 3.2 Problem 2
rovaniemi_data = rovaniemi_data.set_index("YR--MODAHRMN")
daily_averages_rovaniemi = rovaniemi_data.resample("1D").mean()
daily_averages_rovaniemi["CELSIUS_max"] = rovaniemi_data.resample("1D")["CELSIUS"].max()
daily_averages_rovaniemi["CELSIUS_min"] = rovaniemi_data.resample("1D")["CELSIUS"].min()
daily_averages_rovaniemi.head()

USAF       TEMP   MAX   MIN   CELSIUS  CELSIUS_max  \
YR--MODAHRMN                                                          
2017-05-01    28450.0  35.958333  39.0  30.5  2.180556            7   
2017-05-02    28450.0  38.180556  43.5  35.5  3.402778            7   
2017-05-03    28450.0  35.788732  38.0  35.0  2.112676            4   
2017-05-04    28450.0  39.902778  42.5  32.5  4.388889            9   
2017-05-05    28450.0  44.430556  51.0  40.0  6.916667           12   

              CELSIUS_min  
YR--MODAHRMN               
2017-05-01             -1  
2017-05-02              1  
2017-05-03             -1  
2017-05-04             -1  
2017-05-05              1

In [15]:
kumpula_data = kumpula_data.set_index("YR--MODAHRMN")
daily_averages_kumpula = kumpula_data.resample("1D").mean()
daily_averages_kumpula["CELSIUS_max"] = kumpula_data.resample("1D")["CELSIUS"].max()
daily_averages_kumpula["CELSIUS_min"] = kumpula_data.resample("1D")["CELSIUS"].min()
daily_averages_kumpula.head()

USAF       TEMP   MAX   MIN    CELSIUS  CELSIUS_max  \
YR--MODAHRMN                                                           
2017-05-01    29980.0  45.541667  49.0  39.0   7.625000           12   
2017-05-02    29980.0  49.458333  55.5  41.5   9.750000           16   
2017-05-03    29980.0  48.583333  55.0  42.5   9.208333           13   
2017-05-04    29980.0  44.208333  51.5  40.5   6.666667           11   
2017-05-05    29980.0  50.583333  54.5  40.0  10.250000           17   

              CELSIUS_min  
YR--MODAHRMN               
2017-05-01              3  
2017-05-02              2  
2017-05-03              4  
2017-05-04              3  
2017-05-05              2

## All Is Done!